src/lib/proof-system/cache.ts

```diff
@@ -0,0 +1,140 @@
+
+/**
+ * Interface for storing and retrieving values, for caching.
+ * `read()` and `write()` can just throw errors on failure.
+ */
+type Cache = {
+  /**
+   * Read a value from the cache.
+   *
+   * @param header A small header to identify what is read from the cache.
+   */
+  read(header: CacheHeader): Uint8Array | undefined;
+  /**
+   * Write a value to the cache.
+   *
+   * @param header A small header to identify what is written to the cache. This will be used by `read()` to retrieve the data.
+   * @param value The value to write to the cache, as a byte array.
+   */
+  write(header: CacheHeader, value: Uint8Array): void;
+  /**
+   * Indicates whether the cache is writable.
+   */
+  canWrite: boolean;
+};
+
+const cacheHeaderVersion = 0.1;
+
+type CommonHeader = {
+  /**
+   * Header version to avoid parsing incompatible headers.
+   */
+  version: number;
+  /**
+   * An identifier that is persistent even as versions of the data change. Safe to use as a file path.
+   */
+  persistentId: string;
+  /**
+   * A unique identifier for the data to be read. Safe to use as a file path.
+   */
+  uniqueId: string;
+  /**
+   * Specifies whether the data to be read is a utf8-encoded string or raw binary data. This was added
+   * because node's `fs.readFileSync` returns garbage when reading string files without specifying the encoding.
+   */
+  dataType: 'string' | 'bytes';
+};
+type StepKeyHeader<Kind> = {
+  kind: Kind;
+  programName: string;
+  methodName: string;
+  methodIndex: number;
+  hash: string;
+};
+type WrapKeyHeader<Kind> = { kind: Kind; programName: string; hash: string };
+
+/**
+ * A header that is passed to the caching layer, to support richer caching strategies.
+ *
+ * Both `uniqueId` and `programId` can safely be used as a file path.
+ */
+type CacheHeader = (
+  | StepKeyHeader<'step-pk'>
+  | StepKeyHeader<'step-vk'>
+  | WrapKeyHeader<'wrap-pk'>
+  | WrapKeyHeader<'wrap-vk'>
+) &
+  CommonHeader;
+
+const None: Cache = {
+  read() {
+    throw Error('not available');
+  },
+  write() {
+    throw Error('not available');
+  },
+  canWrite: false,
+};
+
+const FileSystem = (cacheDirectory: string): Cache => ({
+  read({ persistentId, uniqueId, dataType }) {
+    if (jsEnvironment !== 'node') throw Error('file system not available');
+
+    // read current uniqueId, return data if it matches
+    let currentId = readFileSync(
+      resolve(cacheDirectory, `${persistentId}.header`),
+      'utf8'
+    );
+    if (currentId !== uniqueId) return undefined;
+
+    if (dataType === 'string') {
+      let string = readFileSync(resolve(cacheDirectory, persistentId), 'utf8');
+      return new TextEncoder().encode(string);
+    } else {
+      let buffer = readFileSync(resolve(cacheDirectory, persistentId));
+      return new Uint8Array(buffer.buffer);
+    }
+  },
+  write({ persistentId, uniqueId, dataType }, data) {
+    if (jsEnvironment !== 'node') throw Error('file system not available');
```

Comment:
Same here for error message? Not sure if it's too verbose but I think giving too much info is always better than not enough

Reply:
this error message will never be read by a user! when `read()` or `write()` throw an error this is just treated as "caching wasn't successful", i.e. it fails silently. this makes it easier to write the caching logic (but arguably a bit harder to debug it..)